hypothesis-python/tests/conjecture/test_engine.py

```diff
@@ -1024,7 +1024,10 @@ def test(data):
 
         runner = ConjectureRunner(test, settings=SMALL_COUNT_SETTINGS)
         runner.run()
-        assert runner.call_count == 256
+        # hits TooHard in generate_novel_prefix. ideally we would generate all
+        # 256 integers in exactly 256 calls here, but in practice we bail out
+        # early.
+        assert runner.call_count >= 200
```

Comment:
Ah, I don't mind the TooHard exception in principle but "can saturate eight bits" doesn't feel like it should be too hard for us?

I'm also concerned that the `exit_reason` statistic is going to say "nothing left to do" and I'd like to add a new "gave up on generating additional unique examples" reason - to avoid misleading users.

Reply:
We can tune the cutoff point of 150 here if desired. I had originally set it to this admittedly quite low value to accompany returning a non-novel prefix instead of giving up entirely, where you pay this overhead in *every* hard draw instead of just one. This approach turned out to be flaky, of course. 

Now that we give up when hitting this limit, I think we can get away with setting it much higher — maybe 500? Though even there this test (as seen on master) may flake.

Saturating n bits requires `2**n` calls on average for the last bit while rejection sampling, so while we could reasonably saturate 8 bits with a TooHard cutoff of e.g. 500, we can't handle too many more bits before I get worried about performance in cases where we really should be bailing.

Reply:
Hmmm, could we use an approach like `UniqueSampledSet` where we use our knowledge of the remaining options to pick one and then write it back to the buffer?  That'd make it more efficient if it works.

Reply:
Is the suggestion here to compute the set of possible children of a node (which is strictly harder/more powerful than computing the max children of that node), then draw one of the not-yet-drawn ones at random?

I think this is the best way forward, and I think you're right that it can be done now, as opposed to the suggestions I made in the source comment. It could also be used in place of our "unbiased" trick, which is similarly trying to avoid re-draws. 

My concern would be in getting this code correct, as enumerating the children is more difficult than counting them. I'm happy to put in that effort up-front in this pr, but it'll probably take me a day or two 🙂

Reply:
Yep.  Or more specifically, try ~ten times to draw one at random (so it's cheap if possible), and if that fails do a maybe-expensive fallback of computing ~100 available options and choosing one of them.  Then next time we try to explore this node, try ~10 again, else top up the cache of available options and repeat.

I think the effort is worth it to avoid a regression here, and maybe enumerating floats isn't quite as bad as you're thinking - the (bits xor sign-bit) representation is lossless and sufficient for ordering, so you can treat as an integer range and then cast back to float.  If NaNs are allowed, just ensure that you use the all-ones-mantissa signaling NaN (2<<64 - 1) as a bound.  

And I just wouldn't worry about the enumeration order distorting things beyond maybe leaving a comment; the regime where random sampling fails ~ten times _and_ we don't come within ~100 samples of exhausting the branch seems pretty unlikely to me.

(idle speculation, in future we might build a smallcheck-like backend using such techniques?  DFS != small though)

Reply:
There were some subtleties (when are there not?), but I've added this mostly as-stated: 9040090000faff39bacf38dd0f940cd9fa2f853a. FWIW I agree with the approach and tradeoffs here. `@given(st.text("a"))` is a case where you do easily fail n > 10 times before coming close to exhausting the search space, which is infinite. But I'm fine with giving pseudo-random output in this case, given that anything else is kind of insanity, lest we try thousand-element-long strings early on. 

Wanted to get this out in case you had any design critiques. Still a few failing tests and test coverage to address, but I'm happy with the direction.

Reply:
Have addressed the failures and coverage here.